src/backend/joanie/api.py

```diff
@@ -0,0 +1,126 @@
+
+from joanie.core import exceptions, models
+
+from . import serializers
+
+
+def get_user_from_token(request):
+    """
+    Decode token from request to get username.
+    Return user found with username or create a new one.
+
+    Args:
+        request: Request
+
+    Return:
+        the User object
+    """
+    authorization_header = request.headers.get("Authorization")
+    access_token = authorization_header.split(" ")[1]  # Bearer XXXXX
+    claims = jwt.decode(
+        access_token,
+        getattr(settings, "JWT_PRIVATE_SIGNING_KEY"),
+        algorithms=getattr(settings, "JWT_ALGORITHM"),
+    )
+    return models.User.objects.get_or_create(username=claims["username"])[0]
+
+
+class OrdersAccessPermission(permissions.BasePermission):
+    """
+    Define a permission to allow access to all orders available or to create an order for a user
+    """
+
+    def has_permission(self, request, view):
+        authorization_header = request.headers.get("Authorization")
+        if not authorization_header:
+            return False
+        access_token = authorization_header.split(" ")[1]  # Bearer XXXXX
+        try:
+            claims = jwt.decode(
+                access_token,
+                getattr(settings, "JWT_PRIVATE_SIGNING_KEY"),
+                getattr(settings, "JWT_ALGORITHM"),
+            )
+            return all(claim in claims for claim in ("iat", "exp"))
+        except Exception:  # pylint: disable=W0703
+            return False
+
+
+class CourseProductsAvailableListView(views.APIView):
+    """
+    For a course, return all products available with its course runs
+    """
+
+    def get(self, request, code):
+        """Return a list of all products available for a course"""
+        course = get_object_or_404(models.Course, code=code)
+        return Response(serializers.ProductSerializer(course.products, many=True).data)
+
+
+class OrderPagination(pagination.PageNumberPagination):
+    """Order pagination to display no more than 100 orders per page"""
+
+    ordering = "-created_on"
+    page_size = 100
+
+
+class OrdersView(generics.ListAPIView):
+    """
+    API view allows to get all orders or create a new one with lms enrollments for a user.
+
+    GET /api/orders/
+        Return list of all orders for a user with pagination
+
+    POST /api/orders/ with expected data:
+        - id: product uid
+        - resource_links: list of resource_links of all course runs selected
+        Return new order just created
+    """
+
+    serializer_class = serializers.OrderSerializer
+    pagination_class = OrderPagination
+    permission_classes = [OrdersAccessPermission]
+
+    def get_queryset(self):
+        user = get_user_from_token(self.request)
+        return (
+            user.orders.all()
+            .select_related("owner", "product")
+            .prefetch_related("enrollments__course_run")
+            # FIXME: add "__positions" is useless, one request for each course run is executed
```

Comment:
I don't think we need to optimize this for now as order queries will never be massive :man_shrugging: